pex/bootstrap.py

```diff
@@ -0,0 +1,76 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import os
+
+
+class Bootstrap(object):
+  """Supports introspection of the PEX bootstrap code."""
+
+  _INSTANCE = None
+
+  @classmethod
+  def locate(cls):
+    """Locates the active PEX bootstrap.
+
+    :rtype: :class:`Bootstrap`
+    """
+    if cls._INSTANCE is None:
+      bootstrap_path = __file__
+      module_import_path = __name__.split('.')
+
+      # For example, our __file__ might be requests.pex/.bootstrap/pex/bootstrap.pyc and our import
+      # path pex.bootstrap; so we walk back through all the module components of our import path to
+      # find the base sys.path entry where we were found (requests.pex/.bootstrap in this example).
+      for _ in module_import_path:
+        bootstrap_path = os.path.dirname(bootstrap_path)
+
+      cls._INSTANCE = cls(location=bootstrap_path)
+    return cls._INSTANCE
+
+  def __init__(self, location):
+    self._location = location
+
+  def demote(self):
+    """Demote the bootstrap code to the end of the `sys.path` so it is found last.
+
+    :return: The list of un-imported bootstrap modules.
+    :rtype: list of :class:`types.ModuleType`
+    """
+    import sys  # Grab a hold of `sys` early since we'll be un-importing our module in this process.
```

Comment:
If I understand correctly, this won't work as you intend if you import at the module level, correct? Or is it here to be closer to the code its used with?

Reply:
The former. You get very strange errors where sys turns up `None` if you leave the imports at the level of the module, the module gets popped from `sys.modules` and then the Bootstrap class is later executed.